sched/jobDefSerializer_test.go

```diff
@@ -0,0 +1,249 @@
+			// compare the orig and generated task definitions
+		} else if ok, msg := origJobDef.Equal(*newJobDef); !ok {
+			panic(fmt.Sprintf("fail: task definitions are not equal: %s\n", msg))
+		} else {
+			fmt.Printf(fmt.Sprintf("ok=%t\n", ok))
+		}
+	}
+}
+
+func makeSampleJobDefintion() JobDefinition {
+	jobDef := JobDefinition{}
+	jobDef.Tasks = make(map[string]TaskDefinition)
+	jobDef.JobType = "jobTypeVal"
+	taskDefinition := TaskDefinition{}
+	taskDefinition.SnapshotId = "snapshotIdVal"
+	tx, _ := time.ParseDuration("+3s")
+	taskDefinition.Timeout = tx
+	envVars := make(map[string]string)
+	taskDefinition.EnvVars = envVars
+	envVars["var1"] = "var1Value"
+	envVars["var2"] = "var2Value"
+	args := []string{"arg1", "arg2"}
+	taskDefinition.Argv = args
+	jobDef.Tasks["task0"] = taskDefinition
+	//Print(jobDef)  -I enable this for debugging
+
+	return jobDef
+}
+
+
+func (this *JobDefinition) Equal(actualJobDef JobDefinition) (bool, string) {
+
+	expectedMap := (*this).Tasks
+	actualMap := actualJobDef.Tasks
+
+
+	if len(expectedMap) != len(actualMap) {
+		return false, fmt.Sprintf("task Definitions maps two different lengths: expected %d, got %d", len(expectedMap), len(actualMap))
+	}
+
+	for taskName, _ := range expectedMap {
+		def1Task, _ := expectedMap[taskName]
+		def2Task, foundTask := actualMap[taskName];
+		if !foundTask {
+			return false, "actual taskDef doesn't contain the expected entry for key:" + taskName
+		}
+		if argvOk, msg := stringSliceCompare(def1Task.Argv, def2Task.Argv); !argvOk {
+			return false, "Argv entries are different:" + msg
+		}
+		if envOk, msg := stringMapEqual(def1Task.EnvVars, def2Task.EnvVars); !envOk {
+			return false, "EnvVars entries are different:" + msg
+		}
+		if def1Task.Timeout.String() != def2Task.Timeout.String() {
+			fmt.Printf(fmt.Sprintf("expected timeout: %s, actual timeout: %s", def1Task.Timeout.String(), def2Task.Timeout.String()))
+			return false, fmt.Sprintf("Timeout values differ, expected %s, got %s", def1Task.Timeout.String(), def2Task.Timeout.String())
+		}
+		if def1Task.SnapshotId != def2Task.SnapshotId {
+			return false, fmt.Sprintf("Snapshot ids differ, expected %s, got %s", def1Task.SnapshotId, def2Task.SnapshotId)
+		}
+	}
+
+	return true, ""
+}
+
+func stringMapEqual(expectedMap, actualMap map[string]string) (bool, string) {
+	if len(expectedMap) != len(actualMap) {
+		return false, fmt.Sprintf("The 2 maps are different lengths: expected %d, got %d", len(expectedMap), len(actualMap))
+	}
+
+	for taskName, entry1 := range expectedMap {
+		if entry2, ok := actualMap[taskName]; !ok || entry1 != entry2 {
+			return false, fmt.Sprintf("The map entries (expected %s, got %s) for %s have different values.", expectedMap[taskName], actualMap[taskName], taskName)
+		}
+	}
+
+	return true, ""
+}
+
+func stringSliceCompare(expectedSlice, actualSlice []string) (bool, string) {
```

Comment:
good thought - unfortunately DeepEqual() returns false negative when slices and maps are empty